File: apps/supervisor-web/src/components/markdownHeuristics.ts
const BLOCK_MARKDOWN_PATTERNS = [
  /^(?: {0,3})#{1,6}\s+\S/m,
  /^(?: {0,3})>{1,}\s*\S/m,
  /^(?: {0,3})(?:[-+*]|\d{1,9}[.)])\s+(?:\[[ xX]\]\s+)?\S/m,
  /^(?: {0,3})(?:```|~~~)/m,
  /^(?: {0,3})(?:[-*_]\s*){3,}$/m,
];
const TABLE_MARKDOWN_PATTERN =
  /^(?:\|?[^|\n]+\|[^|\n]+(?:\|[^|\n]+)*\|?\s*\n\|?\s*:?-{3,}:?\s*(?:\|\s*:?-{3,}:?\s*)+\|?\s*$)/m;
const INLINE_LINK_PATTERN = /!?\[[^\]\n]+\]\([^)]+\)/;
const INLINE_CODE_PATTERN = /`[^`\n]+`/;
const STRONG_EMPHASIS_PATTERN = /(?:\*\*[^*\n]+\*\*|__[^_\n]+__)/;
const EMPHASIS_PATTERN = /(^|[^\w])(?:\*[^*\n]+\*|_[^_\n]+_)(?=[^\w]|$)/;
const STRIKETHROUGH_PATTERN = /~~[^~\n]+~~/;

export function hasLikelyMarkdownSyntax(text: string) {
  const trimmed = text.trim();
  if (!trimmed) {
    return false;
  }

  if (
    BLOCK_MARKDOWN_PATTERNS.some((pattern) => pattern.test(trimmed)) ||
    TABLE_MARKDOWN_PATTERN.test(trimmed)
  ) {
    return true;
  }

  if (!/[`[\]*_~!]/.test(trimmed)) {
    return false;
  }

  return (
    INLINE_LINK_PATTERN.test(trimmed) ||
    INLINE_CODE_PATTERN.test(trimmed) ||
    STRONG_EMPHASIS_PATTERN.test(trimmed) ||
    EMPHASIS_PATTERN.test(trimmed) ||
    STRIKETHROUGH_PATTERN.test(trimmed)
  );
}

File: apps/android/app/src/main/java/com/remotecodex/android/ui/presentation/MarkdownHeuristics.kt
package com.remotecodex.android.ui.presentation

private val blockMarkdownPatterns = listOf(
    Regex("(?m)^(?: {0,3})#{1,6}\\s+\\S"),
    Regex("(?m)^(?: {0,3})>{1,}\\s*\\S"),
    Regex("(?m)^(?: {0,3})(?:[-+*]|\\d{1,9}[.)])\\s+(?:\\[[ xX]\\]\\s+)?\\S"),
    Regex("(?m)^(?: {0,3})(?:```|~~~)"),
    Regex("(?m)^(?: {0,3})(?:[-*_]\\s*){3,}$"),
)

private val tableMarkdownPattern = Regex(
    "(?m)^(?:\\|?[^|\\n]+\\|[^|\\n]+(?:\\|[^|\\n]+)*\\|?\\s*\\n\\|?\\s*:?-{3,}:?\\s*(?:\\|\\s*:?-{3,}:?\\s*)+\\|?\\s*$)",
)
private val inlineLinkPattern = Regex("!?\\[[^\\]\\n]+\\]\\([^)]+\\)")
private val inlineCodePattern = Regex("`[^`\\n]+`")
private val strongEmphasisPattern = Regex("(?:\\*\\*[^*\\n]+\\*\\*|__[^_\\n]+__)")
private val emphasisPattern = Regex("(^|[^\\w])(?:\\*[^*\\n]+\\*|_[^_\\n]+_)(?=[^\\w]|$)")
private val strikethroughPattern = Regex("~~[^~\\n]+~~")

fun hasLikelyMarkdownSyntax(text: String): Boolean {
    val trimmed = text.trim()
    if (trimmed.isEmpty()) {
        return false
    }

    if (
        blockMarkdownPatterns.any { it.containsMatchIn(trimmed) } ||
        tableMarkdownPattern.containsMatchIn(trimmed)
    ) {
        return true
    }

    if (!Regex("[`\\[\\]*_~!]").containsMatchIn(trimmed)) {
        return false
    }

    return inlineLinkPattern.containsMatchIn(trimmed) ||
        inlineCodePattern.containsMatchIn(trimmed) ||
        strongEmphasisPattern.containsMatchIn(trimmed) ||
        emphasisPattern.containsMatchIn(trimmed) ||
        strikethroughPattern.containsMatchIn(trimmed)
}
